File: src/genius/mohammad/portdetective/PortScanner.java
package genius.mohammad.portdetective;

import java.io.InputStream;
import java.net.Socket;

public class PortScanner {
	int progress = 0;	 
	int currentPort = -2;
	String output = "No open ports found...";
	boolean running = true;
	boolean timedOut = false;
	PortDescriptor pd;
	int realPort;
	
	public PortScanner(InputStream stream) {		
		pd = new PortDescriptor(stream);
	}
	
	public void scan() {
		Runnable r = new Runnable() {
			public void run() {
				running = true;
				String ip = RAM.ip;
				int port1 = RAM.port1;
				int port2 = RAM.port2;
				for(int port = port1; port < port2 + 1; port++) {
					if(!timedOut) {						
						realPort = port;
						boolean open = false;
						try {
							Socket s = new Socket(ip, port);
							open = true;
							s.close();
						} catch (Exception e) { }
						if(open) {
							if(output.equals("No open ports found...")) {
								output = "\nPort " + port + " is open.\n" + pd.describePort(port)+"\n";
							}else{
								output += "\nPort " + port + " is open.\n" + pd.describePort(port)+"\n";
							}
						}
						progress = (int)(((port-port1)/(1.0*(port2-port1)) * 100.0) + 0.5);
						currentPort = port;
					}else{
						break;
					}
				}
				if(!timedOut) {
					currentPort = -1;
				}
				running = false;
			}
		};
		Thread t = new Thread(r);
		t.start();
	}	
	
	public void timeOut() {
		timedOut = true;
		running = false;
		currentPort = -2;
	}
}

File: app/src/main/java/genius/mohammad/portdetective/PortScanner.kt
package genius.mohammad.portdetective

import android.text.Html
import android.text.Spanned

import java.io.InputStream
import java.net.Socket

class PortScanner(stream: InputStream) {

    internal var progress = 0
    internal var currentPort = -2
    private var consoleOutput: String? = null
    internal var running = true
    private var timedOut = false
    private val portDescriptor: PortDescriptor = PortDescriptor(stream)

    fun scan(ip: String, portStart: Int, portEnd: Int) {
        stop()
        consoleOutput = "No open ports found..."
        val r = Runnable {
            running = true
            for (port in portStart until portEnd + 1) {
                if (!timedOut && running) {
                    var open = false
                    try {
                        val s = Socket(ip, port)
                        open = true
                        s.close()
                    } catch (e: Exception) {
                    }

                    if (open) {
                        if (consoleOutput == "No open ports found...")
                            consoleOutput = ""
                        consoleOutput += (if (consoleOutput!!.isEmpty()) "" else "<br><br>") + "<b>Port " + port + "</b> is open.<br>" + portDescriptor.getDescription(port)
                    }
                    progress = ((port - portStart) / (1.0 * (portEnd - portStart)) * 100.0 + 0.5).toInt()
                    currentPort = port
                } else {
                    break
                }
            }
            if (!timedOut) {
                currentPort = -1
            }
            running = false
        }
        val t = Thread(r)
        t.start()
    }

    fun getOutput(): Spanned {
        return Html.fromHtml(consoleOutput!!.trim { it <= ' ' })
    }

    fun timeOut() {
        timedOut = true
        running = false
        currentPort = -2
    }

    fun stop() {
        running = false
    }
}
